### ISSUE TITLE
Automatically convert function redefinitions to overloaded functions

### ISSUE THREAD

**Issue by capr** [zdevito/terra#336]
Hello,

Currently, defining a function or method twice results in an error. Could we have it so that it overloads the function that it duplicates instead? (it would still break if the function prototypes are the same).

Rationale: the syntax for adding overloaded functions is very verbose and it makes it hard to refactor functions between overloaded and different-name.

Thanks!

**Comment by elliottslaughter** [comment]
This is how it used to work, it was explicitly changed with rationale described here: https://github.com/zdevito/terra/releases/tag/release-2016-03-25 (under Separate Overloaded and non-Overloaded Functions)

I admit the syntax isn't pretty, maybe that can be cleaned up somehow but the original motivation for separating overloaded/non-overloaded functions still seems applicable to me.

**Comment by capr** [comment]
Hi Elliott, thanks for answering.

What's a REPL ? :) Seriously though, I do use overloads a lot (probably a habit from Lua where any arg can be optional), and I find them pretty frustrating to declare (eg. see https://github.com/luapower/dynarray) mostly because the syntax sugar for methods doesn't work with them and refactoring from a single-signature to a double-signature function and viceversa is a pain, but also because I find it to be an unnecessary burden since the intention (in a module at least) is always to overload, almost never to redefine. So I do hope you'll reconsider this decision :) Maybe enable the new behavior only in the REPL? :D As for saving to .o files and calling from Lua, as long as the function wasn't intentionally overloaded it should still be callable, exportable, etc., no?

**Comment by OvermindDL1** [comment]
As a user, I quite prefer differently-typed redefinitions to be an error by default, I wouldn't want it to transparently overload.  Now if overloading could be enabled on a function name by function name basis with a simple, say, `my_function.overloadable(true)` or so then let differently-typed 'redefinitions' overload that might work, but even that seems unsafe.  I prefer explicit syntax, maybe slightly shorter than what is now, but I definitely prefer explicitness, it catches my programming faults a lot faster.  ^.^

**Comment by capr** [comment]
@OvermindDL1 `my_function.overloadable(true)` would be great.

**Comment by capr** [comment]
Btw, macros have the same loss-of-sugar problem, so why not make macro a keyword too? i.e. instead of writing this:
```
list.methods.next_index = macro(function(self, i) return `self.links(i).next end)
```
I should be able to write this:
```
macro list:next_index(i) return `self.links(i) end
```

**Comment by capr** [comment]
I tried making a proxy table for the struct that would auto-overload function declarations but it doesn't work because it seems that `terra t.f() end` is not exactly syntax sugar for `t.f = terra() end`  and `terra t:f()` is not exactly syntax sugar for `t.methods.f = terra(self: &t) end`, i.e. the duplicate-definition check happens earlier at the parsing stage. If it _were_ syntax sugar, I could make something like this hack: 

```
function autooverload(t)
	local function decl(_, k, v)
		local mt = t.methods[k] and t.methods
			or t.metamethods[k] and t.metamethods
		if mt then
			local m = mt[k]
			if terralib.type(m) == 'terrafunction'
				or terralib.type(m) == 'terramacro'
			then
				mt[k] = terralib.overloadedfunction(k, {m, v})
			elseif terralib.type(m) == 'overloadedterrafunction' then
				mt[k]:adddefinition(v)
			elseif m == nil then
				mt[k] = v
			else
				t[k] = v
			end
		else
			t.methods[k] = v
		end
	end
	return setmetatable({type = t}, {__index = t, __newindex = decl})
end
```

and use it like this:

```
local function make_T()
	local struct T {x: int, y: int}
	T = autooverload(T)
	
	terra T:f() end    -- but I have to use T.f = terra(self: &T) end instead
	terra T:f(i: int) end

	return T.type
end

terra f()
	var t: make_T()
	t:f()
	t:f(5)
end
```

**Comment by elliottslaughter** [comment]
I think there's no way to make this work as syntax sugar without using a shadow table of some sort, because in Lua `__index` and `__newindex` only triggers if the entry is missing completely. (By the way, I think the same goes for your prototype above.)

Usually people work around this by doing something like:

```
setmetatable({__real_values = {}}, {__index = ..., __newindex = ...})
```

Where `__real_values` is the shadow table I described above. E.g. this is how you have do it if you want to implement your own map-like interface in Lua. The issue with doing this is that it then breaks `rawget` and `rawset`, which people could be relying on today and therefore would be a breaking change.

It may be that first-class support would be possible, but off the top of my head the issue I see with it is that functions and overloaded functions are two different kinds of objects, so you're really trying to do the equivalent of desugaring:

```
my_function:overloadable(true)
```

into:

```
local old_function = my_function
my_function = terralib.overloadedfunction("my_function")
my_function:adddefinition(old_function)
```

And I just don't see a way to do that inside a function call. Keep in mind that `my_function` could be a global variable or a local variable, so there is no consistent way to find it even if you knew its name, and there's also no way to know how many other variables could be aliased with it, etc.

The only way I see at the moment to hack this is to reuse the object itself, but replace its guts with the contents of the new overload. E.g.:

```
local old_function = {}
for k, v in pairs(my_function) do
  old_function[k] = v
end
setmetatable(old_function, getmetatable(my_function))
local new_function = terralib.overloadedfunction("my_function")
for k, _ in pairs(old_function) do
  rawset(my_function, k, nil)
end
for k, v in pairs(new_function) do
  rawset(my_function, k, v)
end
setmetatable(my_function, getmetatable(new_function))
my_function:adddefinition(old_function)
```

Which would probably get the job done, but... wow, is this ugly!

**Comment by capr** [comment]
> By the way, I think the same goes for your prototype above.

Nope, that is indeed a proxy table (shadow table), no keys are ever added to it so it should hit `__newindex` every time. Except the duplicate-def check happens _earlier_ so it doesn't have a chance to hit `__newindex` the second time and that's why the hack doesn't work.

I agree that `overloaddable()` called on the function object itself wouldn't work, which is why I suggest that the auto-overloading behavior should only be available (and enabled by default) for structs and it should be a feature _of the struct itself_, applied when actually adding fields to `S.methods` and `S.metamethods` and not earlier as it is now. And remove duplication detection altogether because it is applied inconsistently anyway (eg. it triggers on `local terra f() end` but not on `local f = terra() end`) and because I don't think that it finds any bugs eg. Lua doesn't have the dupe-check and I never once in hundreds of KLOCs ever had a duplicate def by mistake, seriously. Incidentally, this would also solve the REPL issue. What do you think?

**Comment by elliottslaughter** [comment]
Would it be sufficient to just remove the duplicate detection? Then you can customize the behavior of structs to work like you want, without needing this to be a global change for all Terra users.

I agree that I haven't personally had issues with accidentally redefining functions, but I experience a lot of pain from Lua's permissiveness in general. So I'd be happier if there were a way to keep the duplicate check, at least by default. Is there a way to do the check so that it could be customized, like Terra's `strict` checks today?

**Comment by capr** [comment]
@elliottslaughter on a second thought, no, it wouldn't be sufficient, because the desugaring process needs to know the type of `self`, in fact it will probably hit this error: "expected a struct but found table when attempting to add method blah" :( 

So I guess a new metamethod for "new method definition" is probably needed, where you could hook in the duplicate-check or the auto-overload logic.
